crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -2026,28 +2026,58 @@ impl IndexerReader {
     }
 
     fn get_total_supply(&self, coin_struct: StructTag) -> Result<Supply, IndexerError> {
+        if let Some(supply) = self.get_treasury_cap_total_supply(&coin_struct)? {
+            return Ok(supply);
+        }
+        if let Some(supply) = self.get_coin_manager_total_supply(&coin_struct)? {
+            return Ok(supply);
+        }
+        Err(IndexerError::Generic(format!(
+            "Cannot find treasury cap or coin manager for coin type: {}",
+            coin_struct.to_canonical_string(/* with_prefix */ true)
+        )))
+    }
+
+    fn get_treasury_cap_total_supply(
+        &self,
+        coin_struct: &StructTag,
+    ) -> Result<Option<Supply>, IndexerError> {
         let package_id = coin_struct.address.into();
-        let treasury_cap_type =
-            TreasuryCap::type_(coin_struct).to_canonical_string(/* with_prefix */ true);
-        let treasury_cap_obj_id = self
+        let treasury_cap_type = TreasuryCap::type_(coin_struct.clone())
+            .to_canonical_string(/* with_prefix */ true);
+        let cache_key = format!("{}{}", package_id, treasury_cap_type);
+
+        let mut cache = self
             .package_obj_type_cache
             .lock()
-            .unwrap()
-            .cache_get_or_set_with(format!("{}{}", package_id, treasury_cap_type), || {
-                get_single_obj_id_from_package_publish(self, package_id, treasury_cap_type.clone())
-                    .unwrap()
-            })
-            .ok_or(IndexerError::Generic(format!(
-                "Cannot find treasury cap for type {}",
-                treasury_cap_type
-            )))?;
-        let treasury_cap_obj_object =
-            self.get_object(&treasury_cap_obj_id, None)?
-                .ok_or(IndexerError::Generic(format!(
-                    "Cannot find treasury cap object with id {}",
-                    treasury_cap_obj_id
-                )))?;
-        Ok(TreasuryCap::try_from(treasury_cap_obj_object)?.total_supply)
+            .inspect_err(|e| tracing::error!("package_obj_type_cache poisoned: {:?}", e))
+            .map_err(|_| IndexerError::Generic("failed to acquire cache lock".into()))?;
+
+        let maybe_id = if let Some(&id) = cache.cache_get(&cache_key) {
+            id
+        } else {
+            let fetched = get_single_obj_id_from_package_publish(
```

Comment:
It would be more efficient to use `get_singleton_object` that takes advantage of the `objects_package_module_name_full_type` index, similarly to how the `CoinManager` is retrieved currently.

The method used here first queries for the transaction that created the treasure cap, to lookup its id in the transaction effects and then do an additional query to get the object based on the id.